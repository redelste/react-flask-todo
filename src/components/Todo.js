import React, { useRef } from 'react';
import { updateTodo } from '../utils/todoCalls'
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardMedia from "@material-ui/core/CardMedia";
import CardContent from "@material-ui/core/CardContent";
import Divider from "@material-ui/core/Divider";
import Typography from "@material-ui/core/Typography";


const useStyles = makeStyles({
    root: {
        flexGrow: 1,
    },
    card: {
        maxWidth: 300,
        margin: "auto",
        transition: "0.3s",
        boxShadow: "0 8px 40px -12px rgba(0,0,0,0.3)",
        "&:hover": {
            boxShadow: "0 16px 70px -12.125px rgba(0,0,0,0.3)"
        },
        height: "100%"
    },
    media: {
        paddingTop: "56.25%"
    },
    content: {
        textAlign: "left",
        padding: 3
    },
    divider: {
        margin: `10px`
    },
    heading: {
        fontWeight: "bold"
    },
    subheading: {
        lineHeight: 1.8
    },

});


const Todo = ({ todo, todos, setTodos }) => {
    const updateNameInput = useRef(null);
    const updateDescriptionInput = useRef(null);
    const classes = useStyles();

    const handleSubmit = (e, updateNameInput, updateDescriptionInput) => {
        e.preventDefault();
        updateTodo(todo.id, updateNameInput.current.value, updateDescriptionInput.current.value, (data) => {
            // state management --> need to refactor for redux
            const filteredTodos = todos.filter(filterTodo => filterTodo.id !== todo.id);
            const updatedTodo = {
                ...todo,
                name: updateNameInput.current.value || todo.name,
                description: updateDescriptionInput.current.value || todo.description
            }
            setTodos([
                ...filteredTodos,
                updatedTodo
            ])
            if (updateNameInput.current.value !== "" || updateNameInput.current.value !== null) {
                updateNameInput.current.value = "";
            }
            if (updateDescriptionInput.current.value !== "" || updateDescriptionInput.current.value !== null) {
                updateDescriptionInput.current.value = "";
            }
        })

    }

    return <div key={todo.id} className={classes}>
        <Card className={classes.card}>
            <CardMedia
                className={classes.media}
                image={
                    "https://image.freepik.com/free-photo/river-foggy-mountains-landscape_1204-511.jpg"
                }
            />
            <CardContent className={classes.content}>
                <Typography
                    className={"MuiTypography--heading"}
                    variant={"h6"}
                    gutterBottom
                >
                    {todo.name}
                </Typography>
                <Divider className={classes.divider} light />
                <Typography
                    className={"MuiTypography--heading"}
                    variant={"caption"}
                    gutterBottom
                >
                    {todo.datecreated}
                </Typography>
                <Typography
                    className={"MuiTypography--subheading"}
                    variant={"caption"}
                >
                    <Divider className={classes.divider} light />
                    {todo.description}
                    <form onSubmit={(e) => handleSubmit(e, updateNameInput, updateDescriptionInput)}>
                        <button type="submit">
                            Update
                        </button>
                        <input ref={updateNameInput} type="text" name="newName"></input>
                        <input ref={updateDescriptionInput} type="text" name="newDescription"></input>
                    </form>
                    <button onClick={() => {
                        const requestOptions = {
                            method: 'DELETE',
                            headers: { 'Content-Type': 'application/json' },
                            body: JSON.stringify({ id: todo.id })
                        };
                        fetch(`/api/delete`, requestOptions)
                            .then(response => response.json())
                            .then(data => {
                                // This does the same thing as above, except we only want the resultant list.
                                // If the element is equal todo.id, we don't want to keep it. (It's the delete function.)
                                // We do this because we cannot actually mutate todos (it's immutable). So we create a new list "rest", 
                                // which represents the "rest of the todos".
                                const rest = todos.filter(filterTodo => filterTodo.id !== todo.id)
                                setTodos(rest)
                            });
                    }}> Delete</button>
                </Typography>
            </CardContent>
        </Card>
    </div>

}


export default Todo;