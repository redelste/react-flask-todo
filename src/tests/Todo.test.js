import React from 'react';
import  Todo  from "../components/Todo";
import renderer from 'react-test-renderer';

let todo = {datecreated:"2020-08-06T13:56:00", description:"testing", id: 399, name: "a test to "}


let todos =[
    {datecreated:"2020-08-06T13:56:00", description:"testing", id: 399, name: "a test test the jesting"},
    {datecreated:"2020-08-06T14:29:00", description:"testing2", id: 400, name: "a test to test testing"},
    {datecreated:"2020-08-06T15:10:00", description:"testing3", id: 401, name: "a test to jest the jesting"}
]
test('Each todo renders correctly', ()=>{
    const component = renderer.create(<Todo todo={todo} todos={todos}/>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

//  Should be updated in the future to handle axios.
// describe('testing the api functionality through fetch simulation', () =>{
//     it('should add a todo when invoked', ()=>{
//         const wrapper = shallow(<Todo todo={}/>)
//     })
// })