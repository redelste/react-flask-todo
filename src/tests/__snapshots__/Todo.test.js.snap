// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Each todo renders correctly 1`] = `
<div
  className={
    Object {
      "card": "makeStyles-card-2",
      "content": "makeStyles-content-4",
      "divider": "makeStyles-divider-5",
      "heading": "makeStyles-heading-6",
      "media": "makeStyles-media-3",
      "root": "makeStyles-root-1",
      "subheading": "makeStyles-subheading-7",
    }
  }
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root makeStyles-media-3"
      style={
        Object {
          "backgroundImage": "url(\\"https://image.freepik.com/free-photo/river-foggy-mountains-landscape_1204-511.jpg\\")",
        }
      }
    />
    <div
      className="MuiCardContent-root makeStyles-content-4"
    >
      <h6
        className="MuiTypography-root MuiTypography--heading MuiTypography-h6 MuiTypography-gutterBottom"
      >
        a test to 
      </h6>
      <hr
        className="MuiDivider-root makeStyles-divider-5 MuiDivider-light"
      />
      <span
        className="MuiTypography-root MuiTypography--heading MuiTypography-caption MuiTypography-gutterBottom"
      >
        2020-08-06T13:56:00
      </span>
      <span
        className="MuiTypography-root MuiTypography--subheading MuiTypography-caption"
      >
        <hr
          className="MuiDivider-root makeStyles-divider-5 MuiDivider-light"
        />
        testing
        <form
          onSubmit={[Function]}
        >
          <button
            type="submit"
          >
            Update
          </button>
          <input
            name="newName"
            type="text"
          />
          <input
            name="newDescription"
            type="text"
          />
        </form>
        <button
          onClick={[Function]}
        >
           Delete
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`TodoList renders correctly 1`] = `
<div
  className={
    Object {
      "card": "makeStyles-card-2",
      "content": "makeStyles-content-4",
      "divider": "makeStyles-divider-5",
      "heading": "makeStyles-heading-6",
      "media": "makeStyles-media-3",
      "root": "makeStyles-root-1",
      "subheading": "makeStyles-subheading-7",
    }
  }
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root makeStyles-media-3"
      style={
        Object {
          "backgroundImage": "url(\\"https://image.freepik.com/free-photo/river-foggy-mountains-landscape_1204-511.jpg\\")",
        }
      }
    />
    <div
      className="MuiCardContent-root makeStyles-content-4"
    >
      <h6
        className="MuiTypography-root MuiTypography--heading MuiTypography-h6 MuiTypography-gutterBottom"
      >
        a test to 
      </h6>
      <hr
        className="MuiDivider-root makeStyles-divider-5 MuiDivider-light"
      />
      <span
        className="MuiTypography-root MuiTypography--heading MuiTypography-caption MuiTypography-gutterBottom"
      >
        2020-08-06T13:56:00
      </span>
      <span
        className="MuiTypography-root MuiTypography--subheading MuiTypography-caption"
      >
        <hr
          className="MuiDivider-root makeStyles-divider-5 MuiDivider-light"
        />
        testing
        <form
          onSubmit={[Function]}
        >
          <button
            type="submit"
          >
            Update
          </button>
          <input
            name="newName"
            type="text"
          />
          <input
            name="newDescription"
            type="text"
          />
        </form>
        <button
          onClick={[Function]}
        >
           Delete
        </button>
      </span>
    </div>
  </div>
</div>
`;
